engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheReadActor.scala

```diff
@@ -1,32 +1,69 @@
 package cromwell.engine.workflow.lifecycle.execution.callcaching
 
-import akka.actor.{Actor, ActorLogging, Props}
-import cromwell.database.CromwellDatabase
-import cromwell.engine.workflow.lifecycle.execution.{CacheResultLookupFailure, CacheResultMatchesForHashes}
+import akka.actor.{Actor, ActorLogging, ActorRef, Props}
 import cromwell.engine.workflow.lifecycle.execution.callcaching.EngineJobHashingActor.CallCacheHashes
 
 import scala.concurrent.ExecutionContext
-import scala.util.{Failure, Success}
+import akka.pattern.pipe
+import cromwell.core.callcaching.HashResult
+import cromwell.database.sql.MetaInfoId
+import cromwell.engine.workflow.lifecycle.execution.callcaching.CallCacheReadActor._
 
-object CallCacheReadActor {
-  def props(callCacheHashes: CallCacheHashes): Props = {
-    Props(new CallCacheReadActor(callCacheHashes))
-  }
-}
+/**
+  * Queues up work sent to it because its receive is non-blocking.
+  *
+  * Would be nice if instead there was a pull- rather than push-based mailbox but I can't find one...
+  */
+class CallCacheReadActor(cache: CallCache) extends Actor with ActorLogging {
+
+  implicit val ec: ExecutionContext = context.dispatcher
 
-class CallCacheReadActor(callCacheHashes: CallCacheHashes) extends Actor with ActorLogging {
-  {
-    implicit val ec: ExecutionContext = context.dispatcher
+  var requestQueue: List[RequestTuple] = List.empty
+  var currentRequester: Option[ActorRef] = None
 
-    val replyTo = context.parent
-    val cache = new CallCache(CromwellDatabase.databaseInterface)
-    cache.fetchMetaInfoIdsMatchingHashes(callCacheHashes) onComplete {
-      case Success(s) => replyTo ! CacheResultMatchesForHashes(callCacheHashes.hashes, s)
-      case Failure(t) => replyTo ! CacheResultLookupFailure(t)
+  private def runRequest(callCacheHashes: CallCacheHashes) = {
+    val response = cache.fetchMetaInfoIdsMatchingHashes(callCacheHashes) map {
+      case s => CacheResultMatchesForHashes(callCacheHashes.hashes, s)
+    } recover {
+      case t => CacheResultLookupFailure(t)
     }
+
+    response.pipeTo(self)
+  }
+
+  private def cycleRequestQueue() = requestQueue match {
+    case RequestTuple(sndr, request) :: tail =>
+      currentRequester = Option(sndr)
+      requestQueue = tail
+      runRequest(request)
+    case Nil =>
+      currentRequester = None
+  }
+
+  private def receiveNewRequest(callCacheHashes: CallCacheHashes) = currentRequester match {
+    case Some(x) => requestQueue :+= RequestTuple(sender, callCacheHashes)
+    case None =>
+      currentRequester = Some(sender)
+      runRequest(callCacheHashes)
   }
 
   override def receive: Receive = {
+    case CacheLookupRequest(callCacheHashes) => receiveNewRequest(callCacheHashes)
+    case r: CallCacheReadActorResponse =>
+      currentRequester foreach { _ ! r }
```

Comment:
just so I understand this correctly, the CCRA will process a new request once it has been able to respond to its current request? and the cycleQueue is a way to hold the requests in the order they came, so its easy to take work when its time?